File: ts/project/.sdk/tm/swift/Sources/ProjectNameSDK/feature/AuditFeature.swift
// Audit trail. Emits a structured record for every operation - who (actor),
// what (entity + op), the outcome, and a correlation id - suitable for
// compliance logging. Records accumulate on the feature (bounded by `max`,
// default 1000) and, when a `sink` callback is supplied, are also pushed to
// it (e.g. to forward to a SIEM). The actor is the per-call ctrl actor,
// falling back to the options `actor`, then "anonymous". Each operation is
// audited exactly once (the per-context marker in ctx.out prevents a
// preDone + preUnexpected double-log). Timestamps use the injectable `now`
// clock so tests stay deterministic.

import Foundation

public final class AuditFeature: BaseFeature {
  private var client: ProjectNameSDK?
  private var options: VMap?
  private var seq = 0

  // Activity tracking (mirrors the ts client._audit record).
  public var records: [VMap] = []

  private static let seenKey = "audit_seen"

  public override init() {
    super.init()
    version = "0.0.1"
    name = "audit"
    active = true
  }

  public override func initFeature(_ ctx: Context, _ options: VMap) {
    client = ctx.client
    self.options = options
    active = foptBool(options, "active", false)
    seq = 0
  }

  public override func preDone(_ ctx: Context) {
    // Outcome reflects the actual result; a non-2xx reaches preDone
    // before the pipeline errors.
    var outcome = "error"
    if ctx.result != nil && ctx.result!.ok && ctx.result!.err == nil {
      outcome = "ok"
    }
    emit(ctx, outcome)
  }

  public override func preUnexpected(_ ctx: Context) {
    emit(ctx, "error")
  }

  private func emit(_ ctx: Context, _ outcome: String) {
    if !active {
      return
    }

    // One record per operation (preDone + a following preUnexpected on a
    // failure must not double-log).
    if let seen = ctx.out[AuditFeature.seenKey] as? Bool, seen {
      return
    }
    ctx.out[AuditFeature.seenKey] = true

    seq += 1

    var actor = "anonymous"
    let optActor = foptStr(options, "actor", "")
    if optActor != "" {
      actor = optActor
    }
    if ctx.ctrl.actor != "" {
      actor = ctx.ctrl.actor
    }

    let entity = ctx.op?.entity ?? "_"
    let opname = ctx.op?.name ?? "_"

    let record = VMap()
    record.entries["seq"] = .int(Int64(seq))
    record.entries["ts"] = .int(foptNow(options)())
    record.entries["actor"] = .string(actor)
    record.entries["entity"] = .string(entity)
    record.entries["op"] = .string(opname)
    record.entries["outcome"] = .string(outcome)
    record.entries["correlationId"] = .string(ctx.id)
    if let result = ctx.result {
      record.entries["status"] = .int(Int64(result.status))
    }

    records.append(record)
    let max = foptInt(options, "max", 1000)
    while records.count > max {
      records.removeFirst()
    }

    if let sink = gp(options, "sink").asNative as? (VMap) -> Void {
      sink(record)
    }
  }
}

File: ts/project/.sdk/tm/ts/src/feature/audit/AuditFeature.ts

import type { Context, FeatureOptions } from '../../types'
import type { ProjectNameSDK } from '../../ProjectNameSDK'

import { BaseFeature } from '../base/BaseFeature'


class AuditFeature extends BaseFeature {
  version = '0.0.1'
  name = 'audit'
  active = true

  _client?: ProjectNameSDK
  _options: any = {}
  _seq = 0
  _seen: WeakSet<object> = new WeakSet()


  init(ctx: Context, options: FeatureOptions): void | Promise<any> {
    this._client = ctx.client
    this._options = options || {}
    this.active = (options as any).active
    this._seq = 0
    this._seen = new WeakSet()

    const client: any = this._client
    if (null == client._audit) {
      client._audit = { records: [] }
    }
  }


  PreDone(this: any, ctx: any) {
    // Outcome reflects the actual result; a non-2xx reaches PreDone before
    // the pipeline throws.
    this._emit(ctx, (ctx.result && false !== ctx.result.ok && null == ctx.result.err) ? 'ok' : 'error')
  }


  PreUnexpected(this: any, ctx: any) {
    this._emit(ctx, 'error')
  }


  _emit(this: any, ctx: any, outcome: string) {
    // One record per operation (PreDone + a following PreUnexpected on a
    // non-2xx must not double-log).
    if (this._seen.has(ctx)) {
      return
    }
    this._seen.add(ctx)
    this._seq++
    const record = {
      seq: this._seq,
      ts: this._now(),
      actor: (ctx.ctrl && ctx.ctrl.actor) || this._options.actor || 'anonymous',
      entity: (ctx.op && ctx.op.entity) || '_',
      op: (ctx.op && ctx.op.name) || '_',
      outcome,
      status: ctx.result && ctx.result.status,
      correlationId: ctx.id,
    }

    const client: any = this._client
    const recs = client._audit.records
    recs.push(record)
    const max = null == this._options.max ? 1000 : this._options.max
    while (recs.length > max) {
      recs.shift()
    }

    const sink = this._options.sink
    if ('function' === typeof sink) {
      try { sink(record) } catch (_e) { }
    }
  }


  _now(this: any): number {
    const now = this._options.now
    if ('function' === typeof now) {
      return now()
    }
    return Date.now()
  }
}


export {
  AuditFeature
}
